src/networking.c

```diff
@@ -2151,11 +2151,18 @@ void readQueryFromClient(connection *conn) {
         /* Note that the 'remaining' variable may be zero in some edge case,
          * for example once we resume a blocked client after CLIENT PAUSE. */
         if (remaining > 0 && remaining < readlen) readlen = remaining;
+
+        /* Non-greedy to make room for query buffer when bulklen > BIG_ARG. */
+        greedy = 0;
     }
 
     qblen = sdslen(c->querybuf);
     if (c->querybuf_peak < qblen) c->querybuf_peak = qblen;
-    c->querybuf = sdsMakeRoomFor(c->querybuf, readlen);
+    if (!greedy || sdsalloc(c->querybuf) < PROTO_IOBUF_LEN) {
+        c->querybuf = sdsMakeRoomForNonGreedy(c->querybuf, readlen);
+    } else {
+        c->querybuf = sdsMakeRoomFor(c->querybuf, readlen);
```

Comment:
@sundb also note that in case `greedy` is 1 it means that we hope to read as much as possible from the socket. In that case, I think, we should pass `sdsavail` and not `readlen` to the socket read:
```suggestion
        c->querybuf = sdsMakeRoomFor(c->querybuf, readlen);
        readlen = sdsavail(c->querybuf);
```

Reply:
@yoav-steinberg I'm not sure if greed is about reading as much data as possible, the query buffer has already broken 20k (under jemalloc) and since it's already over 20k, let it be greedy.
@oranagra WDYT?

Reply:
i agree with Yoav.
we need to realize that we have two different things that we can be greedy or non-greedy about:
1. memory allocations to avoid repeated reallocs of gradual, incremental growth.
2. read system calls. we normally wanna ask to read for as much as we can (all the memory we allocated), so that in total we'll have less system calls (each one will just return what the OS already has, it's non-blocking anyway).

the case were we don't wanna be greedy (in **both** malloc and read aspects) is the case of the a BIG_ARG, which we later wanna copy (re-purpose) as is into argv, instead of doing a memcpy.

another place were we don't wanna be greedy, is the initial allocation of the default query buffer, we want to start the greediness only above a certain size, in order to avoid the conflict with the resizing cron.

Reply:
I was thinking that if use sdsavalid to read, a connection expanded to 64k would handle a lot more commands than an unexpanded one, but the connection is not fair.

Reply:
yeah, we're not trying to be fair anyway.. so we rather issue less system calls and overall process more commands per second.

Reply:
@oranagra @yoav-steinberg It's dangerous to use ```sdsavail``` to read.
32bit
1) Expand querybuf to over 2.1g.
2) returns of ```connRead``` will likely overflow. The return value of ```connRead``` should be ```ssize_t```, not the current ```int```.

Reply:
>     1. Expand querybuf to over 2.1g.
@sundb I can't see why this change has more of a potential of causing an issue on 32bit than the original code. In both cases we keep expanding the query buff by calling `sdsMakeRoomFor` so it might eventually overflow on a 32bit machine.
> 
>     2. returns of `connRead` will likely overflow. The return value of `connRead` should be `ssize_t`, not the current `int`.
This is a bug in any case. We should change fix connRead to return `ssize_t`.

Reply:
@yoav-steinberg Yes, sdsMakeRoomFor will overflow and trigger the assertion in case of constant expansion.
But before it overflows, ```connRead``` will cause truncation.
In any case, fix ```connRead``` to return ssize_t.

Reply:
`connRead` can be fixed, but note that this is a theoretical issue that can never really happen.
this method reads from a socket or a pipe, it reads what's available in the OS buffers, it'll never be more than a few megs.
@sundb i think your last push should be reverted, i think readlen is better be size_t, and i do think we wanna use sdsavail